lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h

```diff
@@ -214,6 +225,18 @@ struct FunctionSignatureTransformDescriptor {
   /// simply passes it through.
   void addThunkArgument(ArgumentDescriptor &AD, SILBuilder &Builder,
                         SILBasicBlock *BB, SmallVectorImpl<SILValue> &NewArgs);
+
+  /// Whether specializing the function will result in a thunk with the same 
+  /// signature as the original function that calls through to the specialized
+  /// function.
+  ///
+  /// Such a thunk is necessary if there is (or could be) code that calls the
+  /// function which we are unable to specialize to match the function's
+  /// specialization.
+  bool willSpecializationIntroduceThunk() {
+    return !hasOnlyDirectInModuleCallers ||
+           hasPublicVisibility(OriginalFunction->getLinkage());
```

Comment:
I am not sure if hasPublicVisibility is the right check here. Is that a c function? If so, does this properly handle WMO vs non-WMO?

Reply:
Fixed.